designs/simultaneous-edit/README.md

```diff
@@ -0,0 +1,121 @@
+
+![synchronize-actions.png](/Users/nisiyama/src/Curry/Git/designs/designs/simultaneous-edit/images/synchronize-actions.png)
+
+- We may need to synchronize client state at some point by sending refresh request to all client.
+
+- Whether to use sync mode can be specified in `settings.js`.
+
+### Method2: Locking Tab
+
+#### 2-a: Requirements
+
+Lock the flow edit target of client A in order to suppress or notify the edit by another client B.
+
+#### 2-b: Issues
+
+- What to lock to edit?
+
+- What restrictions are placed on locked targets?
+
+#### 2-c: Implementation
+
+- Allow each client to be able to get exclusive editing rights (*lock*) for each tab.
+
+- Place a button to lock the tab on the settings panel of each tab.
+  
+  ![lock-button.png](/Users/nisiyama/src/Curry/Git/designs/designs/simultaneous-edit/images/lock-button.png)
+
+  The lock button expected to have the following states:
+
+| Tab State | Button Label | Description                   |
+| --------- | ------------ | ----------------------------- |
+| Free      | Lock         | Tab is not locked             |
+| Owned     | Unlock       | Tab is locked by this client  |
+| Locked    | Locked       | Tab is locked by other client |
+
+![states.png](/Users/nisiyama/src/Curry/Git/designs/designs/simultaneous-edit/images/states.png)
+
+    
+
+- Tab lock state is managed by the following manner:
+  
+  - If a tab is locked by another client, the tab cannot be edited (**read only**).
+    The tab lock request is sent to the runtime, which manages and arbitrates the state and broadcast the result to other clients.
+  
+  - If a client that holds a lock loses connection to the runtime, the runtime unlocks the tabs that the client is locking after a period of time.  At this point, the runtime also notifies this to other clients.
+  
+  - If the disconnect client reconnects, then
+    
+    - If the tab has not changed, regain the lock.
+    
+    - If the tab has changed, perform the conventional node merge process
+       (*the same issues of flow conflict resolution remain in this case*)
+
+- To make each tab editable by separate clients, it is desirable to divide the flow file into tab-by-tabs.
+
+- Whether to use lock mode can be specified in `settings.js`.
+
+## Discussion
+
+- The problem with method 2 is that conflicts may occur when locks are not acquired and a flow is edited by multiple clients.  So, there still exist the same issues with current Node-RED implementation when conflicts occur.
+
+- Compared with method 2, method 1 has to serialize and broadcast the edit actions to all connected clients. This may cause a performance concern when the number of clients increases.  
+  
+  However, this method seems to be more intuitive for users.  So, I would like to recommend this method as a means for simultaneous flow editing.
```

Comment:
Just to remove any doubt, by 'this method' do you mean method 1 or 2?

Reply:
Sorry for the ambiguous description. It means method 1.